### PR TITLE
fixed error in  factorial.py

### PR DESCRIPTION
**Pull request by SandersLin**
### **Describe your change:**
`factorial(11)` returns` index out of range`. To implement memoization an auxiliary function is required.

* [ ] Add an algorithm?
* [x] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### **Checklist:**
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [ ] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [ ] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [ ] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: dynamic_programming/factorial.py [hunk 1]
```diff
@@ -12,26 +10,26 @@ def factorial(num):
...
 
     if num < 0:
         return "Number should not be negative."
+
+    return factorial_aux(num, result)
+
+
+def factorial_aux(num, result):
+
+    if num == 0 or num == 1:
```
Comment (cclauss) [inline dynamic_programming/factorial.py:23]:
```suggestion
    if num in (0, 1):
```

Comment (swapnanildutta) [reply dynamic_programming/factorial.py:23]: This makes more sense.

**Comment by pawarashish564** [comment]
we can use `math.factorial` if required .

File: dynamic_programming/factorial.py [hunk 2]
```diff
@@ -1,37 +1,37 @@
 # Factorial of a number using memoization
-result = [-1] * 10
-result[0] = result[1] = 1
 
 
 def factorial(num):
```
Comment (cclauss) [inline dynamic_programming/factorial.py:4]:
```suggestion
def factorial(num: int) -> int:
```

File: dynamic_programming/factorial.py [hunk 3]
```diff
@@ -1,37 +1,37 @@
...
     """
+    result = [-1] * (num + 1)
 
     if num < 0:
         return "Number should not be negative."
+
+    return factorial_aux(num, result)
+
+
+def factorial_aux(num, result):
```
Comment (cclauss) [inline dynamic_programming/factorial.py:23]: Type hints?  Doctests?

File: dynamic_programming/factorial.py [hunk 4]
```diff
@@ -1,37 +1,37 @@
...
+    if num == 0 or num == 1:
+        return 1
+
     if result[num] != -1:
         return result[num]
     else:
-        result[num] = num * factorial(num - 1)
-        # uncomment the following to see how recalculations are avoided
-        # print(result)
+        result[num] = num * factorial_aux(num - 1, result)
```
Comment (fhlasek) [inline dynamic_programming/factorial.py:31]: Technically, this is more of a "[memoization](https://stackoverflow.com/questions/6184869/what-is-the-difference-between-memoization-and-dynamic-programming#:~:text=Both%20Memoization%20and%20Dynamic%20Programming,solving%20the%20problem%20bottom%2Dup.)" approach. While being perfectly correct, the more appropriate algorithm for this section may look as follows 
```suggestion
    result[0] = 1
    for i in xrange(num):
        result.append((i + 1) * result[i])
    return result[num]
```

Comment (cclauss) [reply dynamic_programming/factorial.py:31]: xrange() was removed from Python on 1/1/2020.

The canonical way to do memoization in Python is with the [__lru_cache__](https://docs.python.org/3/library/functools.html#functools.lru_cache) function decorator.

**Pull request merged**